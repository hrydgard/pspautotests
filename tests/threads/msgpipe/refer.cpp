#include "shared.h"

void testRefer(const char *title, SceUID msgpipe, SceKernelMppInfo *info) {
	int result = sceKernelReferMsgPipeStatus(msgpipe, info);
	if (result == 0) {
		checkpoint(NULL);
		schedf("%s: ", title);
		schedfMsgPipe(info);
	} else {
		checkpoint("%s: Failed (%08x)", title, result);
	}
}

extern "C" int main(int argc, char **argv) {
	SceUID msgpipe = sceKernelCreateMsgPipe("refer", PSP_MEMORY_PARTITION_USER, 0, 0, NULL);
	SceKernelMppInfo info;
	info.size = sizeof(info);

	// Crashes.
	//testRefer("NULL info", msgpipe, NULL);
	testRefer("Normal", msgpipe, &info);

	checkpointNext("Sizes:");
	int sizes[] = {0, 1, 2, 4, 8, 44, 48, 52, 56, 64, 128, 256, 1024, 0xFF00, 0xFF0000, 0x80000001, 0xCCCCCCCC, 0xFFFFFFFF};
	int i, result;
	for (i = 0; i < sizeof(sizes) / sizeof(sizes[0]); ++i) {
		info.size = sizes[i];
		result = sceKernelReferMsgPipeStatus(msgpipe, &info);
		checkpoint("  %08X => %08X (result=%08X)", sizes[i], info.size, result);
	}
	schedf("\n");
	flushschedf();

	sceKernelDeleteMsgPipe(msgpipe);

	info.size = sizeof(info);
	checkpointNext("Objects:");
	testRefer("NULL", 0, &info);
	testRefer("Invalid", 0xDEADBEEF, &info);
	testRefer("Deleted", msgpipe, &info);

	flushschedf();
	return 0;
}