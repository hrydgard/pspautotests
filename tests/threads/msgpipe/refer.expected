[x] Normal: Msgpipe: OK (size=56, name=refer, attr=00000000, buffer=0, free=0, sending=0, receiving=0)

[x] Sizes:
[x]   00000000 => 00000000 (result=00000000)
[x]   00000001 => 00000038 (result=00000000)
[x]   00000002 => 00000038 (result=00000000)
[x]   00000004 => 00000038 (result=00000000)
[x]   00000008 => 00000038 (result=00000000)
[x]   0000002C => 00000038 (result=00000000)
[x]   00000030 => 00000038 (result=00000000)
[x]   00000034 => 00000038 (result=00000000)
[x]   00000038 => 00000038 (result=00000000)
[x]   00000040 => 00000038 (result=00000000)
[x]   00000080 => 00000038 (result=00000000)
[x]   00000100 => 00000038 (result=00000000)
[x]   00000400 => 00000038 (result=00000000)
[x]   0000FF00 => 00000038 (result=00000000)
[x]   00FF0000 => 00000038 (result=00000000)
[x]   80000001 => 00000038 (result=00000000)
[x]   CCCCCCCC => 00000038 (result=00000000)
[x]   FFFFFFFF => 00000038 (result=00000000)

[x] Objects:
[x] NULL: Failed (8002019e)
[x] Invalid: Failed (8002019e)
[x] Deleted: Failed (8002019e)
